### DISCUSSION TITLE
Can we make it easy to add custom workflows?

### DISCUSSION THREAD

**Discussion by SMUsamaShah** [zero01101/openOutpaint#34] [Ideas]
The new history pane is so nice. We can even have floating preview panes and stuff next. 

What can we do so that we can manually make something like this https://ps.reddit.com/r/StableDiffusion/comments/z0gfho/smooth_infinite_zoom_slow/ in OO (openOutpaint)? The linked video was all automatic by a script. Doing it manually should allow quality stuff like the original infinite zoom painting https://zoomquilt.org/

For infinite zoom we need 
 - resize an image 
 - we can remove part of image already
 - a way to save only selected region of the canvas

We also need mask strength because it does work https://github.com/zero01101/openOutpaint/discussions/28 I don't know why other tools have not done this yet. Sometimes you might want to just morph something a little bit instead of complete repaint. This will help with that. Will just need 
- mask strength which can be represented by the mask transparency

**Comment by seijihariki** [comment]
Regarding mask strength I think that if we implement this as a separate feature, at this point, we may as well just implement a separate toolbar for various mask and image editing tools (gradient insertion, border-preserving img2img, etc.).

The main floating window is getting quite crowded after all.

**Reply by seijihariki** [reply]
This is now in the works; should have a working first version by tomorrow

**Reply by seijihariki** [reply]
Will try to finish it tonight, but yesterday night started to make a tool that I believe will be quite useful for a lot of things, 
including image scaling! That could help a lot for this workflow you described.

**Reply by seijihariki** [reply]
Well, took a while, but #49 does quite some groundwork for some interesting tools in the future.

**Reply by SMUsamaShah** [reply]
That is so amazing. The image cropping and resizing moving etc. we are going photoshop 🚀 

How easy do you think it currently is to add something new code wise e.g. if I wanted to add a new tool or extend an existing one? Mostly I am interested in playing with masks (including preview panel) and some basic ui stuff.

**Reply by seijihariki** [reply]
I think it is much easier than before, at least in the "everything is in one place" department. The simplest tool to check is the mask paintbrush, that should give you a good idea of how to implement a simple tool.

Most of the implementation regarding the mask brush is in the paintbrush.js and toolbar.js

**Reply by seijihariki** [reply]
Sorry, it's actually maskbrush.js

For a bit more complex, the dream ones are quite well documented, and stamp.js is okay too. 

I just don't recommend trying to start understanding the code from the select.js, as it is kind of a monstrosity for now, but it may be useful for understanding how to use keyboard and mouse events together, and how a tool can interact with another (stamp.js).

**Comment by zero01101** [comment]
>I just don't recommend trying to start understanding the code from the select.js, as it is kind of a monstrosity for now

lol did you forget the state of this repo before you wonderful folks came along and cleaned up so much of my mess ;)
